training-agent: /mcp-strict no longer rejects unsigned create_media_buy with 401

## Symptom

The integration test \`server/tests/integration/training-agent-strict.test.ts > Training Agent /mcp-strict route > presence-gated enforcement > unsigned create_media_buy on /mcp-strict returns 401 request_signature_required\` is failing on \`main\` (commit a6844988d at time of filing).

```
AssertionError: expected 200 to be 401 // Object.is equality
- Expected: 401
+ Received: 200

❯ tests/integration/training-agent-strict.test.ts:120:26
    118|         idempotency_key: '550e8400-e29b-41d4-a716-446655440000',
    119|       });
    120|       expect(res.status).toBe(401);
```

## Why this matters

The \`/mcp-strict\` route is the grader target for \`adcp grade request-signing\` and the AAO Verified compliance dashboard. It's supposed to enforce \`required_for: ['create_media_buy']\` with presence-gated auth — an unsigned \`create_media_buy\` request must produce \`401 request_signature_required\` so the compliance runner exercises vector 001.

If the route is silently bearer-fall-through-ing on \`create_media_buy\` instead of rejecting, then:

1. Vector 001 (\`request_signature_required\`) never fires against the training agent.
2. The grader can't distinguish a correctly-failing implementation from a silently-permissive one when pointed at \`/mcp-strict\`.
3. Runners using the training agent as a positive control may believe their pipeline is verifying signatures when it isn't.

## Reproduction

\`\`\`
git checkout main
cd server
npx vitest run tests/integration/training-agent-strict.test.ts
\`\`\`

Five tests pass, three fail (two of the three are the test asserting the old \`specialisms: ['signed-requests']\` claim — those will be fixed by #3079; the 401 failure is independent and exists on \`main\` today).

## What this might be

The strict-route handler in \`server/src/training-agent/index.ts\` has a fairly elaborate dance with the framework default flag (\`useFrameworkServer()\`). The presence-gated rejection is supposed to fire from the request-signing pre-transport hook before the bearer authenticator gets a chance, but recent moves (framework default flip, test controller integration, \`requireTokenStrict\` introduction) may have inverted the order or wrapped the rejection in something that swallows it.

A debug-log or breakpoint at the unsigned strict path should show whether:
- The request-signing pre-transport hook is even running on \`/mcp-strict\`
- The 401 is being emitted but then overwritten by a later layer
- The route is just hitting the bearer authenticator and short-circuiting

## Out of scope here

#3079 (signed-requests specialism cleanup) noted this failure but did not fix it — keeping that PR scoped to the deprecation cleanup. This issue tracks the strict-route regression as its own thing.

## References

- Originally added in #2474 (training-agent strict route + presence-gated auth)
- Test file: \`server/tests/integration/training-agent-strict.test.ts:107\`
- Route handler: \`server/src/training-agent/index.ts:533\`

--- a/.changeset/fix-strict-rawbody-resolve-operation.md
+++ b/.changeset/fix-strict-rawbody-resolve-operation.md
@@ -1,0 +1,18 @@
+---
+---
+
+fix(training-agent): restore 401 for unsigned create_media_buy on /mcp-strict
+
+`resolveOperation` in `buildStrictAuthenticator` read `req.rawBody` exclusively and silently returned `undefined` when it was absent. In test harnesses that mount `express.json()` without the production `verify` callback, `rawBody` is never populated, so the `required_for: ['create_media_buy']` gate never fired and unsigned requests authenticated via bearer → 200.
+
+**Two-part fix:**
+
+1. `server/src/training-agent/index.ts` — `buildStrictAuthenticator`'s `resolveOperation` now falls back to `req.body` (already parsed by express.json) when `rawBody` is absent. Safe because `resolveOperation` drives only the `required_for` routing decision, not cryptographic verification — the signing authenticator's own `resolveOperation` deliberately remains `rawBody`-only.
+
+2. `server/src/training-agent/request-signing.ts` — `enforceSigningWhenWebhookAuthPresent` applies the same fallback for the webhook-authentication downgrade-resistance check so a test harness without the `verify` callback still catches `push_notification_config.authentication` payloads.
+
+3. `server/tests/integration/training-agent-strict.test.ts` — `beforeAll` now mirrors production `http.ts` by adding the `verify` callback to `express.json`, populating `req.rawBody` for all test requests.
+
+Non-protocol server change; `--empty` changeset per playbook.
+
+Closes #3080.

--- a/server/src/training-agent/index.ts
+++ b/server/src/training-agent/index.ts
@@ -166,11 +166,17 @@ function buildStrictAuthenticator(): Authenticator | null {
     fallback: bearerAuth,
     requiredFor: STRICT_REQUIRED_FOR,
     resolveOperation: (req) => {
+      // rawBody is populated by the production http.ts `verify` callback.
+      // Fall back to req.body (already-parsed by express.json) when rawBody
+      // is absent — e.g. in test harnesses that skip the verify callback.
+      // Safe here because resolveOperation drives only the required_for
+      // routing decision, not cryptographic verification.
       const raw = (req as { rawBody?: string }).rawBody;
-      if (!raw) return undefined;
       try {
-        const body = JSON.parse(raw) as { method?: string; params?: { name?: string } };
-        if (body.method === 'tools/call' && typeof body.params?.name === 'string') {
+        const body = raw
+          ? JSON.parse(raw) as { method?: string; params?: { name?: string } }
+          : (req as { body?: { method?: string; params?: { name?: string } } }).body;
+        if (body && body.method === 'tools/call' && typeof body.params?.name === 'string') {
           return body.params.name;
         }
       } catch {

--- a/server/src/training-agent/request-signing.ts
+++ b/server/src/training-agent/request-signing.ts
@@ -254,7 +254,11 @@ function subtreeHasWebhookAuthentication(node: unknown, depth = 0): boolean {
 export function enforceSigningWhenWebhookAuthPresent(inner: Authenticator): Authenticator {
   const wrapped: Authenticator = async (req) => {
     if (!requestCarriesSignatureHeader(req.headers)) {
-      const raw = (req as { rawBody?: string }).rawBody;
+      // rawBody from the production http.ts verify callback; fall back to
+      // re-serialising req.body for test harnesses that omit the callback.
+      const rawBody = (req as { rawBody?: string }).rawBody;
+      const bodyForWebhookCheck = (req as { body?: unknown }).body;
+      const raw = rawBody ?? (bodyForWebhookCheck !== undefined ? JSON.stringify(bodyForWebhookCheck) : undefined);
       if (raw && bodyCarriesWebhookAuthentication(raw)) {
         throw new AuthError(
           'Signature required when push_notification_config.authentication is present.',

--- a/server/src/training-agent/request-signing.ts
+++ b/server/src/training-agent/request-signing.ts
@@ -258,7 +258,12 @@ export function enforceSigningWhenWebhookAuthPresent(inner: Authenticator): Auth
       // re-serialising req.body for test harnesses that omit the callback.
       const rawBody = (req as { rawBody?: string }).rawBody;
       const bodyForWebhookCheck = (req as { body?: unknown }).body;
-      const raw = rawBody ?? (bodyForWebhookCheck !== undefined ? JSON.stringify(bodyForWebhookCheck) : undefined);
+      // Guard: JSON.stringify is safe only for plain objects (express.json always
+      // produces one, but a misconfigured parser could yield a Buffer or string).
+      const rawFallback = bodyForWebhookCheck !== null && typeof bodyForWebhookCheck === 'object' && !Array.isArray(bodyForWebhookCheck) && !Buffer.isBuffer(bodyForWebhookCheck)
+        ? JSON.stringify(bodyForWebhookCheck)
+        : undefined;
+      const raw = rawBody ?? rawFallback;
       if (raw && bodyCarriesWebhookAuthentication(raw)) {
         throw new AuthError(
           'Signature required when push_notification_config.authentication is present.',

--- a/server/tests/integration/training-agent-strict.test.ts
+++ b/server/tests/integration/training-agent-strict.test.ts
@@ -77,7 +77,14 @@ describe('Training Agent /mcp-strict route', () => {
 
   beforeAll(() => {
     app = express();
-    app.use(express.json());
+    // Mirror production http.ts: populate req.rawBody via the verify callback
+    // so requireTokenStrict's resolveOperation can identify the tool name and
+    // apply the required_for gate without falling back to req.body.
+    app.use(express.json({
+      verify: (req, _res, buf) => {
+        (req as unknown as { rawBody?: string }).rawBody = buf.toString('utf8');
+      },
+    }));
     app.use('/api/training-agent', createTrainingAgentRouter());
   });
 

